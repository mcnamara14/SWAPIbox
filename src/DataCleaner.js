class DataCleaner {
  fetchData = async (filter) => {
    const peopleData = await fetch(`https://swapi.co/api/${filter}`)
    const response = await peopleData.json()
    const rawData = await response.results

    switch(filter) {
      case 'people':
          return this.cleanPeopleData(rawData)
          break;
      default:
          null
    }

  };



  fetchPeopleData = (people) => {
    const unresolvedPeopleData = people.map(async (person) => {
      const homeworldResponse = await fetch(person.homeworld);
      const parsedHomeworld = await homeworldResponse.json();
      const speciesResponse = await fetch(person.species);
      const parsedSpecies = await speciesResponse.json();

      return {name: person.name, homeworld: parsedHomeworld.name, 
              species: parsedSpecies.name, population: parsedHomeworld.population}
    });

    return Promise.all(unresolvedPeopleData)
  }

  cleanPeopleData = async (rawData) => {
    const cleanPeopleData = await this.fetchPeopleData(rawData)
    
    return cleanPeopleData
  }

}

export default DataCleaner;