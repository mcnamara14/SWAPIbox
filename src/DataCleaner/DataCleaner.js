class DataCleaner {
  fetchData = async (filter) => {
    const fetchedData = await fetch(`https://swapi.co/api/${filter}`);
    const response = await fetchedData.json();
    const rawData = response.results;

    switch (filter) {
      case 'people':
        return this.cleanPeopleData(rawData);
      case 'planets':
        return this.cleanPlanetData(rawData);
      case 'vehicles':
        return this.cleanVehicleData(rawData);
    }
  };

  cleanPeopleData = async (people) => {
    const unresolvedPeopleData = people.map(async (person, index) => {
      const name = person.name;
      const homeworldData = await this.cleanHomeworldData(person.homeworld);
      const homeworldName = homeworldData.name;
      const homeworldPopulation = homeworldData.population;
      const species = await this.cleanSpeciesData(person.species);
      
      return {name, homeworld: homeworldName, 
        species, population: homeworldPopulation, id: `people${index}`};
    });

    return await Promise.all(unresolvedPeopleData);
  };

  cleanHomeworldData = async (homeworld) => {
    const homeworldResponse = await fetch(homeworld);
    const homeworldData = await homeworldResponse.json();
    const name = homeworldData.name;
    const population = homeworldData.population;

    return { name, population };
  }

  cleanSpeciesData = async (species) => {
    const speciesResponse = await fetch(species);
    const speciesData = await speciesResponse.json();
    const speciesName = speciesData.name;

    return speciesName;
  }

  cleanPlanetData = async (planets) => {
    const unresolvedPlanetData = planets.map(async (planet, index) => {
      const residents = await this.cleanResidentsData(planet);
 
      return {name: planet.name, terrain: planet.terrain, 
        population: planet.population, climate: planet.climate, 
        residents,  id: `planets${index}`};
    });

    return await Promise.all(unresolvedPlanetData);
  };

  cleanResidentsData = async (planet) => {
    const unresolvedResidents = planet.residents.map(async (resident) => {
      const residentsResponse = await fetch(resident);
      const parsedResident = await residentsResponse.json();
      const residentName = parsedResident.name;

      return residentName;
    });

    return await Promise.all(unresolvedResidents);
  };

  cleanVehicleData = async (vehicles) => {
    const unresolvedVehiclesData = vehicles.map(async (vehicle, index) => {
      return {name: vehicle.name, model: vehicle.model, 
        class: vehicle.vehicle_class, passengers: vehicle.passengers, 
        id: `vehicles${index}`};
    });

    return await Promise.all(unresolvedVehiclesData);
  }
 
}

export default DataCleaner;