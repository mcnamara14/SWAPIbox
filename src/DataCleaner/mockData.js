export const mockCleanedVehicleData = {
    class: "sail barge",
    id: "vehicles0",
    model: "Modified Luxury Sail Barge",
    name: "Sail barge",
    passengers: "500"
}

export const mockCleanedPeopleData = {
  homeworld: "Tatooine",
  id: "people0",
  name: "Luke Skywalker",
  population: "200000",
  species: "Human"
}

export const mockCleanedPlanetData = {
  climate: "temperate",
  id: "planets0",
  name: "Alderaan",
  residents: "Alderaan",
  population: "2000000000",
  terrain: "grasslands, mountains"
}

export const mockDirtyPlanetData = {
  "count": 61,
  "next": "https://swapi.co/api/planets/?page=2",
  "previous": null,
  "results": [
      {
        "name": "Alderaan",
        "rotation_period": "24",
        "orbital_period": "364",
        "diameter": "12500",
        "climate": "temperate",
        "gravity": "1 standard",
        "terrain": "grasslands, mountains",
        "surface_water": "40",
        "population": "2000000000",
        "residents": [
            "https://swapi.co/api/people/5/",
            "https://swapi.co/api/people/68/",
            "https://swapi.co/api/people/81/"
        ],
        "films": [
            "https://swapi.co/api/films/6/",
            "https://swapi.co/api/films/1/"
        ],
        "created": "2014-12-10T11:35:48.479000Z",
        "edited": "2014-12-20T20:58:18.420000Z",
        "url": "https://swapi.co/api/planets/2/"
      }
    ]
  }

export const mockDirtyPeopleData = {
  "count": 87,
  "next": "https://swapi.co/api/people/?page=2",
  "previous": null,
  "results": [
      {
          "name": "Luke Skywalker",
          "height": "172",
          "mass": "77",
          "hair_color": "blond",
          "skin_color": "fair",
          "eye_color": "blue",
          "birth_year": "19BBY",
          "gender": "male",
          "homeworld": "https://swapi.co/api/planets/1/",
          "films": [
              "https://swapi.co/api/films/2/",
              "https://swapi.co/api/films/6/",
              "https://swapi.co/api/films/3/",
              "https://swapi.co/api/films/1/",
              "https://swapi.co/api/films/7/"
          ],
          "species": [
              "https://swapi.co/api/species/1/"
          ],
          "vehicles": [
              "https://swapi.co/api/vehicles/14/",
              "https://swapi.co/api/vehicles/30/"
          ],
          "starships": [
              "https://swapi.co/api/starships/12/",
              "https://swapi.co/api/starships/22/"
          ],
          "created": "2014-12-09T13:50:51.644000Z",
          "edited": "2014-12-20T21:17:56.891000Z",
          "url": "https://swapi.co/api/people/1/"
      }
  ]
}

export const mockDirtyVehicleData = {
  "count": 39,
  "next": "https://swapi.co/api/vehicles/?page=2",
  "previous": null,
  "results": [
      {
          "name": "Sail barge",
          "model": "Modified Luxury Sail Barge",
          "manufacturer": "Ubrikkian Industries Custom Vehicle Division",
          "cost_in_credits": "285000",
          "length": "30",
          "max_atmosphering_speed": "100",
          "crew": "26",
          "passengers": "500",
          "cargo_capacity": "2000000",
          "consumables": "Live food tanks",
          "vehicle_class": "sail barge",
          "pilots": [],
          "films": [
              "https://swapi.co/api/films/3/"
          ],
          "created": "2014-12-18T10:44:14.217000Z",
          "edited": "2014-12-22T18:21:15.807906Z",
          "url": "https://swapi.co/api/vehicles/24/"
      }
  ]
}

export const mockDirtyHomeworldData = {
  "name": "Tatooine",
  "rotation_period": "23",
  "orbital_period": "304",
  "diameter": "10465",
  "climate": "arid",
  "gravity": "1 standard",
  "terrain": "desert",
  "surface_water": "1",
  "population": "200000",
  "residents": [
      "https://swapi.co/api/people/1/",
      "https://swapi.co/api/people/2/",
      "https://swapi.co/api/people/4/",
      "https://swapi.co/api/people/6/",
      "https://swapi.co/api/people/7/",
      "https://swapi.co/api/people/8/",
      "https://swapi.co/api/people/9/",
      "https://swapi.co/api/people/11/",
      "https://swapi.co/api/people/43/",
      "https://swapi.co/api/people/62/"
  ],
  "films": [
      "https://swapi.co/api/films/5/",
      "https://swapi.co/api/films/4/",
      "https://swapi.co/api/films/6/",
      "https://swapi.co/api/films/3/",
      "https://swapi.co/api/films/1/"
  ],
  "created": "2014-12-09T13:50:49.641000Z",
  "edited": "2014-12-21T20:48:04.175778Z",
  "url": "https://swapi.co/api/planets/1/"
}

export const mockDirtySpeciesData = {
  "name": "Droid",
  "classification": "artificial",
  "designation": "sentient",
  "average_height": "n/a",
  "skin_colors": "n/a",
  "hair_colors": "n/a",
  "eye_colors": "n/a",
  "average_lifespan": "indefinite",
  "homeworld": null,
  "language": "n/a",
  "people": [
      "https://swapi.co/api/people/2/",
      "https://swapi.co/api/people/3/",
      "https://swapi.co/api/people/8/",
      "https://swapi.co/api/people/23/",
      "https://swapi.co/api/people/87/"
  ],
  "films": [
      "https://swapi.co/api/films/2/",
      "https://swapi.co/api/films/7/",
      "https://swapi.co/api/films/5/",
      "https://swapi.co/api/films/4/",
      "https://swapi.co/api/films/6/",
      "https://swapi.co/api/films/3/",
      "https://swapi.co/api/films/1/"
  ],
  "created": "2014-12-10T15:16:16.259000Z",
  "edited": "2015-04-17T06:59:43.869528Z",
  "url": "https://swapi.co/api/species/2/"
}

export const mockDirtyFilmsData = [
    {
        "title": "A New Hope",
        "episode_id": 4,
        "opening_crawl": "It is a period of civil war to the galaxy....",
        "director": "George Lucas",
        "producer": "Gary Kurtz, Rick McCallum",
        "release_date": "1977-05-25",
        "characters": [
            "https://swapi.co/api/people/1/",
            "https://swapi.co/api/people/2/"
        ],
        "planets": [
            "https://swapi.co/api/planets/2/"
        ],
        "starships": [
            "https://swapi.co/api/starships/2/",
            "https://swapi.co/api/starships/3/",
            "https://swapi.co/api/starships/5/"
        ],
        "vehicles": [
            "https://swapi.co/api/vehicles/4/"
        ],
        "species": [
            "https://swapi.co/api/species/5/",
            "https://swapi.co/api/species/3/"
        ],
        "created": "2014-12-10T14:23:31.880000Z",
        "edited": "2015-04-11T09:46:52.774897Z",
        "url": "https://swapi.co/api/films/1/"
    }
]

export const mockDirtyResidentData = {

}