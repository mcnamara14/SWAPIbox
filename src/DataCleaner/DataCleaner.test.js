import React from 'react';
import DataCleaner from './DataCleaner';
import { mockDirtyPeopleData, mockCleanedPeopleData, mockCleanedPlanetData, 
         mockDirtyHomeworldData, mockDirtySpeciesData, mockDirtyPlanetData,
        mockDirtyVehicleData, mockCleanedVehicleData} from './mockData';
import { shallow } from 'enzyme';

describe('Data cleaner', () => {
  let dataCleaner;

  beforeEach(() => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(
        mockDirtyPeopleData
      )
    }))
   
   dataCleaner = new DataCleaner();
  })

  it('should match snapshot', () => {
    expect(DataCleaner).toMatchSnapshot();
  });

  describe('fetchData', () => {
    it('should fetch people data with the corrent url', async () => {
      const filter = 'people';
      const expected = 'https://swapi.co/api/people';

      await dataCleaner.fetchData(filter)

      expect(window.fetch).toHaveBeenCalledWith(expected);
    })

    it('should call cleanPeopleData with the correct arguments based on filter', async () => {
      const filter = 'people';
      const mockData = mockDirtyPeopleData.results;

      dataCleaner.cleanPeopleData = jest.fn();

      await dataCleaner.fetchData(filter)

      expect(dataCleaner.cleanPeopleData).toHaveBeenCalledWith(mockData);
    })

    it('should call cleanPlanetData with the correct arguments based on filter', async () => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve(
          mockDirtyPlanetData
        )
      }))

      const filter = 'planets';
      const mockData = mockDirtyPlanetData.results;

      dataCleaner.cleanPlanetData = jest.fn();

      await dataCleaner.fetchData(filter)

      expect(dataCleaner.cleanPlanetData).toHaveBeenCalledWith(mockData);
    })

    it('should call cleanVehicleData with the correct arguments based on filter', async () => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve(
          mockDirtyVehicleData
        )
      }))

      const filter = 'vehicles';
      const mockData = mockDirtyVehicleData.results;

      dataCleaner.cleanVehicleData = jest.fn();

      await dataCleaner.fetchData(filter)

      expect(dataCleaner.cleanVehicleData).toHaveBeenCalledWith(mockData);
    })

  });

  describe('cleanPeopleData', () => {
    
    beforeEach(() => {
      dataCleaner.cleanHomeworldData = jest.fn().mockImplementation(() => { 
        return {name: 'Tatooine' , population: '200000' }
      });
      dataCleaner.cleanSpeciesData = jest.fn().mockImplementation(() => {
        return 'Human'
      })
    })

    it('should call cleanHomeworldData and cleanSpeciesData with the correct args', async () => {
      const mockData = mockDirtyPeopleData.results
      const expectedHomeworldArg = mockDirtyPeopleData.results[0].homeworld
      const expectedSpeciesArg = mockDirtyPeopleData.results[0].species

      await dataCleaner.cleanPeopleData(mockData)

      expect(dataCleaner.cleanHomeworldData).toHaveBeenCalledWith(expectedHomeworldArg)
      expect(dataCleaner.cleanSpeciesData).toHaveBeenCalledWith(expectedSpeciesArg)
    })
    
    it('should return an array of objects containing clean data', async () => {
      const mockData = mockDirtyPeopleData.results;
      const mockCleanPeopleData = [mockCleanedPeopleData];
      const result = await dataCleaner.cleanPeopleData(mockData)

      expect(result).toEqual(mockCleanPeopleData)
    })

  });

  describe('cleanHomeworldData', async () => {
    beforeEach(() => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve(
          mockDirtyHomeworldData
        )
      }))
    })

    it('should call fetch with the correct argument', async () => {
      const homeworld = mockDirtyPeopleData.results[0].homeworld
      const expected = 'https://swapi.co/api/planets/1/'

      await dataCleaner.cleanHomeworldData(homeworld)

      expect(window.fetch).toHaveBeenCalledWith(expected);
    })

    it('should return an object with the homeworld name and population', async () => {
      const expected = { name: "Tatooine", population: "200000" }
      const homeworld = mockDirtyPeopleData.results[0].homeworld

      const result = await dataCleaner.cleanHomeworldData(homeworld);

      expect(result).toEqual(expected)
    })

  });

  describe('cleanSpeciesData', async () => {
    let species;

    beforeEach(() => {
      const species = mockDirtyPeopleData.results[0].species
      
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve(
          mockDirtySpeciesData
        )
      }));
    })
    
    it('should call fetch with the correct argument', async () => {
      const species = mockDirtyPeopleData.results[0].species[0]
      const expected = 'https://swapi.co/api/species/1/'

      await dataCleaner.cleanSpeciesData(species)

      expect(window.fetch).toHaveBeenCalledWith(expected);
    })

    it('should return an object with the homeworld name and population', async () => {
      const expected = 'Droid'

      const result = await dataCleaner.cleanSpeciesData(species)

      expect(result).toEqual(expected)
    })

  });

  describe('cleanPlanetData', () => {
    
    it('should call cleanResidentsData with correct arg', async () => {
      dataCleaner.cleanResidentsData = jest.fn() 
      const expected = mockDirtyPlanetData
   
      await dataCleaner.cleanPlanetData([mockDirtyPlanetData]);

      expect(dataCleaner.cleanResidentsData).toHaveBeenCalledWith(expected);
    });

    it('should return an array of objects containing clean data', async () => {
      const mockData = mockDirtyPlanetData.results
      const expected = [mockCleanedPlanetData]

      dataCleaner.cleanResidentsData = jest.fn().mockImplementation(() => { 
        return 'Alderaan'
      });

      const result = await dataCleaner.cleanPlanetData(mockData);

      expect(result).toEqual(expected);
    })
  });
  
  describe('cleanResidentsData', () => {

    it('should call fetch with the correct argument', async () => {
      const mockData = mockDirtyPlanetData.results[0]
      const result = dataCleaner.cleanPeopleData = jest.fn();

      await dataCleaner.cleanResidentsData(mockData)

      expect(window.fetch).toHaveBeenCalledTimes(3)
    })

  })

  describe('cleanVehicleData', () => {

    it('should call fetch with the correct argument', async () => {
      const mockData = mockDirtyVehicleData.results
      const mockCleanVehicleData = [mockCleanedVehicleData]

      const result = await dataCleaner.cleanVehicleData(mockData)

      expect(result).toEqual(mockCleanVehicleData)
    })

  })

});