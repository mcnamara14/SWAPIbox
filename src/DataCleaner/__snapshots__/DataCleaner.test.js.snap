// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data cleaner shoul match snapshot 1`] = `[Function]`;
